File: i-payout/recipes/java/spendback.java
package java;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

public class spendback {

  private static final String BASE_URL = "https://merchantapi.testewallet.com/api/v1";
  private static final String USERNAME = "<username>"; // Replace with your actual username
  private static final String PASSWORD = "<password>"; // Replace with your actual password
  private static String authStr = USERNAME + ":" + PASSWORD;
  private static String encodedAuthStr = Base64.getEncoder().encodeToString(authStr.getBytes(StandardCharsets.UTF_8));

  private static String apiToken;
  private static final String BENEFICIARY_TOKEN = "<beneficiaryToken>"; // Replace with actual beneficiary token
  private static final String CURRENCY_CODE = "USD"; // Replace with actual currency code if different

  public static void main(String[] args) throws Exception {
    authenticate();
    checkBalance();
    performSpendback();
  }

  // Authenticate the user and retrieve the API token
  private static void authenticate() throws Exception {
    URL url = new URL(BASE_URL + "/authentication/login");
    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
    conn.setRequestMethod("GET");
    conn.setRequestProperty("Accept", "application/json");
    conn.setRequestProperty("Authorization", "Basic " + encodedAuthStr);
    conn.setRequestProperty("X-MerchantId", USERNAME);

    int responseCode = conn.getResponseCode();
    if (responseCode == HttpURLConnection.HTTP_OK) {
      String response = new String(conn.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
      // Manually parse JSON to extract the token
      String tokenKey = "\"token\":\"";
      int tokenStart = response.indexOf(tokenKey) + tokenKey.length();
      int tokenEnd = response.indexOf("\"", tokenStart);
      apiToken = response.substring(tokenStart, tokenEnd);
    } else {
      throw new RuntimeException("Failed to authenticate: " + responseCode);
    }
    conn.disconnect();
  }

  // Check the balance of the beneficiary
  private static void checkBalance() throws Exception {
    URL url = new URL(BASE_URL + "/beneficiaries/" + BENEFICIARY_TOKEN + "/balances/" + CURRENCY_CODE);
    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
    conn.setRequestMethod("GET");
    conn.setRequestProperty("Authorization", "Bearer " + apiToken);
    conn.setRequestProperty("X-MerchantId", USERNAME);
    conn.setRequestProperty("Accept", "application/json");

    int responseCode = conn.getResponseCode();
    String _response = conn.getResponseMessage();
    if (responseCode == HttpURLConnection.HTTP_OK) {
      String response = new String(conn.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
      // Print the balance data
      System.out.println(response);
    } else {
      throw new RuntimeException("Failed to check balance: " + responseCode);
    }
    conn.disconnect();
  }

  // Perform the spendback operation
  private static void performSpendback() throws Exception {
    URL url = new URL(BASE_URL + "/payins/spendback");
    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
    conn.setRequestMethod("POST");
    conn.setRequestProperty("Authorization", "Bearer " + apiToken);
    conn.setRequestProperty("X-MerchantId", USERNAME);
    conn.setRequestProperty("Accept", "application/json");
    conn.setRequestProperty("Content-Type", "application/*+json");

    // Manually create the JSON string for the request body
    String jsonBody = "{"
        + "\"amount\":100.00," // Replace with the amount you want to spend back
        + "\"currency\":\"USD\"," // Replace with actual currency if different
        + "\"reference\":\"spendback_001\""
        + "}";

    conn.setDoOutput(true);
    try (OutputStream os = conn.getOutputStream()) {
      os.write(jsonBody.getBytes(StandardCharsets.UTF_8));
      os.flush();
    }

    int responseCode = conn.getResponseCode();
    if (responseCode == HttpURLConnection.HTTP_OK) {
      String response = new String(conn.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
      System.out.println(response);
    } else {
      throw new RuntimeException("Failed to perform spendback: " + responseCode);
    }
    conn.disconnect();
  }
}

File: i-payout/recipes/python/spendback.py
import base64
import requests

base_url = "https://merchantapi.testewallet.com/api/v1"

username = "<username>"  # Replace with your actual username
password = "<password>"  # Replace with your actual password
auth_str = f"{username}:{password}"
encoded_auth_str = base64.b64encode(auth_str.encode()).decode()

token_url = f"{base_url}/authentication/login"
headers = {
    "accept": "application/json",
    "authorization": f"Basic {encoded_auth_str}",
    "X-MerchantId": username,
}

response = requests.get(token_url, headers=headers)
api_token = response.json()["data"]["token"]
merchant_id = username

beneficiary_token = "<beneficiaryToken>"  # Replace with actual beneficiary token
currency_code = "USD"  # Replace with actual currency code if different
balance_url = f"{base_url}/beneficiaries/{beneficiary_token}/balances/{currency_code}"

headers = {
    "Authorization": f"Bearer {api_token}",
    "X-MerchantId": merchant_id,
    "accept": "application/json",
}

balance_response = requests.get(balance_url, headers=headers)
balance_data = balance_response.json()

spendback_url = f"{base_url}/payins/spendback"
spendback_body = {
    "amount": 100.00,  # Replace with the amount you want to spend back
    "currency": "USD",  # Replace with actual currency if different
    "reference": "spendback_001",
}

headers = {
    "Authorization": f"Bearer {api_token}",
    "X-MerchantId": merchant_id,
    "accept": "application/json",
    "content-type": "application/*+json",
}

spendback_response = requests.post(spendback_url, headers=headers, json=spendback_body)
spendback_data = spendback_response.json()

print(spendback_data.json())
